gqcp/include/QCModel/HF/UHF.hpp

```diff
@@ -345,6 +347,656 @@ class UHF {
+     *      A'_I(sigma)A(sigma)J(sigma_bar)B(sigma_bar) = (A(sigma)I(sigma)|J(sigma_bar)B(sigma_bar)).
+     * 
+     *  @note Sigma_bar = alpha if sigma is beta and vice versa.
+     * 
+     *  @note The name `Spin-conserved`comes from the article "Constraints and stability in Hartree-Fock theory" by Seeger, R. and Pople J.A. (https://doi.org/10.1063/1.434318).
+     * 
+     *  @param rsq_hamiltonian      The second quantized Hamiltonian, which contains the necessary two-electron operators.
+     *  @param sigma                The spin-sigma component. 
+     *  @param sigma_bar            The spin component, opposite to sigma.
+     * 
+     *  @return The mixed spin sigma component of stability matrix A'.
+     */
+    GQCP::Matrix<Scalar> calculateMixedSpinConservedAComponent(const RSQHamiltonian<Scalar>& rsq_hamiltonian, const Spin sigma, const Spin sigma_bar) const {
+
+        // Sigma and sigma_bar need to be different. If they are the same, the method throws an error.
+        if (sigma == sigma_bar) {
+            throw std::invalid_argument("QCModel::UHF<Scalar>.calculateMixedSpinConservedAComponent(const RSQHamiltonian<Scalar>& rsq_hamiltonian, const Spin sigma, const Spin sigma_bar): The spin 'sigma' and spin 'sigma_bar' arguments are not allowed to be the same.");
+        }
+
+        // Depending on whether we are making the aabb or bbaa A'-component, we need a different component of the orbital_space.
+        const auto orbital_space = this->orbitalSpace();
+        const auto orbital_space_sigma = this->orbitalSpace(sigma);
+        const auto orbital_space_sigma_bar = this->orbitalSpace(sigma_bar);
+
+        // Determine the number of occupied and virtual orbitals for both spin components.
+        const auto n_occ_sigma = orbital_space_sigma.numberOfOrbitals(OccupationType::k_occupied);
+        const auto n_virt_sigma = orbital_space_sigma.numberOfOrbitals(OccupationType::k_virtual);
+
+        const auto n_occ_sigma_bar = orbital_space_sigma_bar.numberOfOrbitals(OccupationType::k_occupied);
+        const auto n_virt_sigma_bar = orbital_space_sigma_bar.numberOfOrbitals(OccupationType::k_virtual);
+
+        // We need the two-electron integrals in mixed MO basis.
+        // The ground state coefficient matrix is obtained from the QCModel.
+        const auto& two_electron_integrals = rsq_hamiltonian.twoElectron();
+
+        // We transform the necessary integrals to a 'MixedUSQTwoElectronOperatorComponent` in order to be able to transform the tensor indices to a mixed basis.
+        const auto g_op = GQCP::ScalarMixedUSQTwoElectronOperatorComponent<Scalar> {two_electron_integrals.parameters()};
+
+        // Transform the first two indices to the sigma basis and the last two indices to the sigma_bar basis.
+        // Return the parameters for later use.
+        const auto g = g_op.transformed(this->expansion(Spin::alpha), Spin::alpha).transformed(this->expansion(Spin::beta), Spin::beta).parameters();
```

Comment:
I've looked at the implementation of all of the components of the stability matrices, and it seems that you're always transforming the 'restricted AO spin-orbitals' to the UHF basis (= your mixed MO basis).

Instead of doing so, maybe I could suggest an alternative approach?
- Let the user set up a `USQHamiltonian` in the basis of the orthonormal UHF spin-orbitals, obtained from a converged UHF calculation.
- Work with that orthonormal USQHamiltonian instead of having to transform the integrals in these methods, because it encapsulates the (orthonormal) two-electron operators `g_aaaa`, `g_aabb`, `g_bbaa` and `g_bbbb`.

Reply:
Wouldn't this make it less clear for the user? Now the stability analysis is completely analogous for all Hartree-Fock methods. If I change this, GHF and RHF will have the same way of performing a stability analysis, with practically the same API's, but UHF will require an extra step, i.e. creating a second hamiltonian. Additionally I believe it can be quite confusing for the user to suddenly have to work with a `USQHamiltonian` while all other calculations in UHF are done using the `RSQHamiltonian`.

Reply:
It was implied that the RHF and GHF stability analyses would also use the `RSQHamiltonian` and `GSQHamiltonian` _in the HF orbital bases_.

I don't agree on your comment that this API can be confusing: a method developer should be aware that stability analysis requires a Hamiltonian in an orthonormal basis and SCF calculations requires the Hamiltonian in the original AO basis.

Reply:
I see your point. I just don't fully see why it is okay for RHF and GHF to implicitly transform their Hamiltonian to an orthonormal basis, but not for UHF. 

Wouldn't it make more sense to treat all methods equally? So, if it is the responsibility of the developer to know when to use a Hamiltonian in the orthonormal basis, wouldn't it be best to apply this change to RHF and GHF as well?

> It was implied that the RHF and GHF stability analyses would also use the RSQHamiltonian and GSQHamiltonian in the HF orbital bases.

Or is that exactly what you meant by this? 😅

Reply:
> Or is that exactly what you meant by this? 😅

Yep! So we're on the same line here.